src/libraries/System.Collections/ref/System.Collections.cs

```diff
@@ -415,6 +415,13 @@ public partial struct Enumerator : System.Collections.Generic.IEnumerator<T>, Sy
             void System.Collections.IEnumerator.Reset() { }
         }
     }
+    public sealed class ReferenceEqualityComparer : System.Collections.Generic.IEqualityComparer<object>, System.Collections.IEqualityComparer
```

Comment:
This should implement `IEqualityComparer<object?>`

Reply:
Why did we choose to hardcode object rather than it being generic with a class constraint?

Reply:
Why wouldn't `IEqualityComparer<object>` be correct? The methods on that interface are already marked with `[AllowNull]` and `[DisallowNull]` as appropriate.

Additionally, the use of a non-generic class means that there can be a single instance that can work for all possible `T`, as the `IEqualityComparer<T>` interface is contravariant. There was no need to introduce a `ReferenceEqualityComparer<T>` generic class; it would just make the JIT do unnecessary work and wouldn't open any new scenarios. Due to the contravariant behavior the below code will work just fine with the non-generic `ReferenceEqualityComparer` type.

```cs
var dict = new Dictionary<string, int>(ReferenceEqualityComparer.Instance); // contravariant on specific T

public static void GenericMethod<T>() where T : class // note constraint for open T
{
    var dict = new Dictionary<T, int>(ReferenceEqualityComparer.Instance);
}
```

Reply:
> as the IEqualityComparer<T> interface is contravariant

Right, I was thinking of `IEquatable<T>`, which is invariant.

> Why wouldn't `IEqualityComparer<object>` be correct?

For the same reason: contravariance.  Let's say I have `class Foo<T> { public Foo(IEqualityComparer<T>) { } }`.  If I have an `IEqualityComparer<object?>`, I can use that with a non-nullable T or a nullable T, but if I have `IEqualityComparer<object>`, I'll get a warning if I try to use that with a nullable T.  Essentially `object?` is the new top type in the nullable reference types world.

Reply:
It sounds like `IEqualityComparer<object>` and `IEqualityComparer<object?>` would have members with identical nullability. If so, why not use `IEqualityComparer<object?>` just to make things as clear as possible up front?

Reply:
I'm not following your response; I'm saying it should be `object?`.

Reply:
Sorry @stephentoub, your comment wasn't visible when I replied under Levi's comment. I fixed it.

Reply:
Ah, I think I get it. Is this what you mean?

```cs
public static void Foo()
{
    Receiver<string>(ReferenceEqualityComparer.Instance);
    Receiver<string?>(ReferenceEqualityComparer.Instance); // warning if ReferenceEqualityComparer is IEC<obj>, not IEC<obj?>
}

private static void Receiver<T>(IEqualityComparer<T> comparer)
{
}
```

Reply:
> Is this what you mean?

Yes.  This is why, for example, StringComparer implements `IEqualityComparer<string?>` rather than `IEqualityComparer<string>`.  This comparer should follow suit.